service/history/workflowTaskHandler.go

```diff
@@ -568,9 +569,7 @@ func (handler *workflowTaskHandlerImpl) handleCommandCompleteWorkflow(
 		return handler.failWorkflowTask(enumspb.WORKFLOW_TASK_FAILED_CAUSE_UNHANDLED_COMMAND, nil)
 	}
 
-	if handler.updateRegistry.HasUndeliveredUpdates() {
-		return handler.failWorkflowTask(enumspb.WORKFLOW_TASK_FAILED_CAUSE_UNHANDLED_UPDATE, nil)
-	}
+	handler.updateRegistry.CancelUpdates(ctx, workflow.WithEffects(handler.effects, handler.mutableState))
```

Comment:
The shutting down of a run seems like something that should be done by a workflow.Context rather than in WFT handling, no? Like maybe the WFT handler calls wfctx.Terminate but the steps that the workflow actually takes should be encapsulated.

Reply:
I tend to agree with Matt. The close of workflow should asynchronously trigger the cleanup of the mutable state or some sort that would close the pending updates.